### ISSUE TITLE
Create tool to test runtime-rs API

### ISSUE THREAD

**Issue by jodh-intel** [kata-containers/kata-containers#5269]
Since Kata switched to the shimv2 architecture, it is very difficult to run the runtime "stand-alone" (without a container manager). Therefore, it makes sense to consider creating a small tool, similar to the [agent-ctl tool](https://github.com/kata-containers/kata-containers/tree/main/src/tools/agent-ctl), that would allow the [`runtime-rs hypervisor` API](https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/hypervisor/README.md) to be tested for the configured hypervisor.

This would support the runtime-rs hypervisor integration work.

See:

- https://github.com/kata-containers/kata-containers/issues/4634
- https://github.com/orgs/kata-containers/projects/33/views/1

**Comment by jodh-intel** [comment]
@pmores, @gkurz - Are either of you interested in working on this? I'm happy to be involved too, but I think one of you mentioned this idea recently?

**Comment by pmores** [comment]
@jodh-intel I'm very interested in a task like this in principle however

- unfortunately I have to be a bit careful not to oversubscribe myself, given the time available (I'm pretty much a novice here and progressing rather slowly)
- perhaps more importantly, I have very little idea how to go about it so senior mentoring would be necessary.

I seem to remember running into numerous obstacles trying to run the runtime-rs binary last winter. I think it required an unpacked container image, several sockets to talk to cri-o and others and what not.  I tried to mock those ad-hoc as I could but didn't get very far.

I can see fundamentally two ways how to approach the tool,
- change the shim binary so that it's more robust in presence of an incomplete environment and only fails once a missing piece of environment is actually needed.  For instance, I imagine a kata pod VM could basically be started without a container image.
- or, on the contrary, leave shim alone and put all responsibility on the tool to create whatever shim needs - give it images, be cri-o and agent at the same time to the shim.

I assume the latter approach would be preferable but I'm not sure if it can be done.  Without being clear on such fundamental architectural decisions, I wouldn't be comfortable to start coding.

**Comment by jodh-intel** [comment]
I'll tal at this...

**Comment by Tim-0731-Hzt** [comment]
Hi @jodh-intel, I also vwould like to be involeved to this.

**Comment by pmores** [comment]
I've been looking into the latter of the two approaches I outlined above.  It's possible to get the binary running by shamelessly faking the arguments along the lines of
`sudo KATA_RUNTIME_BIND_FD=0 ./containerd-shim-kata-v2 -namespace default -address /home/pvl/sock -publish-binary /bin/echo -id xxx`
although it obviously does nothing yet.  I'd still like to understand more about the state that a thus launched shim is in.

The `-namespace` and `-id` arguments look easy enough but I've been wondering about the rest. Is a valid `-publish-binary` necessary to run the shim in a development mode?

Also, I haven't been able to find any information about the purpose of `KATA_RUNTIME_BIND_FD` (googling it returns literally nothing ;-)). Surely I could decipher it from the code and I'll try to do it if necessary but maybe someone has insights to spare me from that?

Lastly, as for the `-address`, I assume that it's the communication interface with containerd/cri-o and if so, the [shim v2](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md) protocol is spoken over that channel.  Would that be correct?  If so then I assume our tool would need to be able to speak (a subset of) that protocol.

**Comment by Tim-0731-Hzt** [comment]
If you are trying to communicating with the `shimv2` via socket in `-address`, so I assume the tool you want to make is like a `containerd` with limited functionality ?

**Comment by pmores** [comment]
@Tim-0731-Hzt I _suspect_ that the socket in argument of `--address` is the one connecting the shim to containerd and so shim v2 ttrpc runs over it but I'm not sure yet. I'll investigate more next week unless someone can give me more information until that.

The log produced by the faked command line (see my previous message) goes like this:
```

{"msg":"start to run","level":"INFO","ts":"2022-10-20T10:21:10.153548643Z","subsystem":"shim","source":"xxx","name":"kata-runtime","pid":"3214813","version":"0.1.0"}
{"msg":"begin to run service","level":"INFO","ts":"2022-10-20T10:21:10.15391717Z","subsystem":"service","name":"kata-runtime","version":"0.1.0","source":"xxx","pid":"3214813"}
{"msg":"wait server message","level":"INFO","ts":"2022-10-20T10:21:10.154236035Z","version":"0.1.0","source":"xxx","pid":"3214813","subsystem":"service","name":"kata-runtime"}
```
which seems to suggest, unless I'm mistaken, that the shim now waits for messages to guide its further activity.

OTOH I seem to remember that when first tried to run the shim back in February (I think) it was able to proceed to container image bundle processing with no messages received.  Maybe this was a feature of an early-stage shim that's not in the current code base anymore, but maybe there's still a way to get the shim to process `configuration.toml`, the bundle etc. and try to run a configured hypervisor even without messing with the containerd shim v2 ttrpc.  That would make things easier.

**Comment by Tim-0731-Hzt** [comment]
@pmores you understanding is correct, `containerd ` is talking to `shim` via this socket using ttrpc.  If you get this log, it means the `servicemanager` already start running.
```rust
info!(sl!(), "begin to run service");
self.start().await.context("start")?;
info!(sl!(), "wait server message");
```
So, it's able to recieve the ttrpc request now.

**Comment by Tim-0731-Hzt** [comment]
There are some default path for `configuration.toml`, so shim will process that without messing with the ttrpc, but the `bundle` is send from the ttrpc

**Comment by jodh-intel** [comment]
@pmores - I'm actually looking at a slightly lower-level tool. What I'm after is a CLI tool that can essentially do exactly what the containerd-shim-kata-v2 binary does, but:

- Have no dependency on containerd or its API.
- Be able to call all the dependent setup functions (hypervisor config, virtiofsd, etc).
- Allow the user to specify which Hypervisor API calls they wish to make (ordering, number of repeated calls, etc).

Basically, something like the `agent-ctl` tool. That tool is different to the Kata runtime ("shim") in that it allows arbitrary agent API calls to be made, whereas the runtime only ever makes the correct sequence of agent API calls.

Let's talk about this more in the meeting next week...

**Comment by pmores** [comment]
@jodh-intel Perhaps I should explain where I'm coming from here so that we can figure out how to share effort.  My next milestone (at least the way I see it now) would be to add a bare minimum of code to `prepare_vm()` and `start_vm()` and have a way to exercise it so that I have a QEMU process running.  For this first stage it's enough to ignore a vast majority of `configuration.toml` and basically only use the path to the hypervisor binary, the path to a rootfs and probably the amount of RAM for the VM (the QEMU default of 128 MB seems too small for most use cases).

This is where I'd like get as soon as possible because I think it would mean that I have a practicable dev cycle to add features and verify and debug them.

This is also why I'm looking into how to mock up as much of the shim runtime environment as necessary for this.  However, if the tool you have in mind would provide an alternative path I'm all for that.  I think I know where I need to get and I'll take any path that leads there.

Do I understand correctly that you're looking into a binary independent of the shim which uses the shim code to handle configuration, hypervisor lifecycle etc. which however relies on a more accessible mechanism to invoke it?

**Comment by pmores** [comment]
@Tim-0731-Hzt Yes you're right.  I think I was confused for a while by the `--bundle` argument that the shim also takes which however seems to serve a different purpose.

**Comment by Tim-0731-Hzt** [comment]
@pmores  `--bundle` in shim argument can also be used to start a container by running the binary `containerd-shim-kata-v2`, and the bundle path directory should include the `config.json `and container rootfs.

**Comment by pmores** [comment]
@Tim-0731-Hzt That's interesting.  According to the shim v2 proto spec, `--bundle` is not meant to be used by the `start` command but it's used by the `delete` command to clean up resources left behind after e.g. containerd loses a connection to a shim.

If it's possible to use `--bundle` for `start` as well that would be a protocol extension I guess?

**Comment by Tim-0731-Hzt** [comment]
@pmores Yes, it is possible, you can refer to the docs in [`runc`](https://github.com/opencontainers/runc) and [`crun`](https://github.com/containers/crun/blob/main/crun.1.md). Normally, the container runtime will implement this.

**Comment by jodh-intel** [comment]
> Do I understand correctly that you're looking into a binary independent of the shim which uses the shim code to handle configuration, hypervisor lifecycle etc. which however relies on a more accessible mechanism to invoke it?

Yes. A tool that only provides a Hypervisor API level access isn't that helpful imho as most of the logic for supporting the different hypervisors will be around handling the configuration.

But the hypervisor configuration would require a tool "just below the level of the shim" (aka full control over the hypervisor and its config, the OCI bundle and the Kata config), but without all the containerd API. However, that's not particularly easy given the fact that the tool would need to break some encapsulation and expose some currently private crate APIs to allow it to call the hypervisor API from a higher level API.

**Comment by pmores** [comment]
I'm wondering if it's even enough to expose Hypervisor implementations as some important stuff is happening in `runtimes/virt_container`, `service` and possibly other upper-layer crates (like cgroups set-up or `virtiofsd` launch which happen in hypervisor-independent code).

**Comment by jodh-intel** [comment]
A few other points:

- @pmores asked about obtaining further details on pre/post conditions for the various Hypervisor API calls.
- Although the runtime-rs codebase supports using an external virtiofs daemon, this code is untested (since Dragonball doesn't need this).
- We (the Hypervisor API implementors atleast) need to do more reading and research for ideas 1 and 2.
- If idea 3 is a possibility now, we could use that functionality to accelerate development. A lower-level tool would still be useful later to test individual APIs.
- We will arrange a follow-up meeting in a week or so to get more of an update on the ideas and progress on implementations.

**Comment by studychao** [comment]
@jodh-intel Thanks!James!

**Comment by pmores** [comment]
Thanks @jodh-intel !  I'm investigating idea 2 to see if I can find a convenient boundary between the containerd-specific outer layer(s) and containerd-independent inner implementation.  If there is one, or if one could be created by a reasonably simple and non-invasive refactor, I'd look into making it possible to replace the containerd-specific outer shell with something that can supply the inner implementation with stuff that it needs (which normally comes from the containerd shim v2 interface) in a way more practicable for development.

In the meantime I'm also interested in idea 3 - could someone give me a bit more specific outline of how that would be done?

Also to clarify my remarks on testing down at the Hypervisor level - the docs have been extended and I believe they are more than adequate, at least for now.  What I had in mind in the call was that the members of the Hypervisor trait might be inherently hard to unit-test as it seems non-trivial to describe in a hypervisor-independent way what the post-conditions for a call like `prepare_vm()` might be. And integration testing might be complicated too due to the interplay of hypervisor-specific functionality in Hypervisor trait implementations and hypervisor-independent code.  This is not to say that it cannot be done, let alone that it wouldn't be useful - just that it seems a bit tricky to me at the moment.

**Comment by pmores** [comment]
So it turns out that while the both highest layers of the shim (`ShimExecutor` and `ServiceManager` in crates `shim` and `service` respectively) are containerd-dependent, the next lower one (`RuntimeHandlerManager` in `runtimes`) is not.  `RuntimeHandlerManager` is driven at OCI level by its `handler_message()` function and is fairly easy to instantiate and use (kudos to runtime-rs authors for the architecture!).  This allowed me to write a little program as follows:

```
use runtimes::RuntimeHandlerManager;
use common::{
    message::Message,
    types::{Request, ContainerConfig},
};
use tokio::{
    sync::mpsc::{channel, Receiver},
};

const MESSAGE_BUFFER_SIZE: usize = 8;

#[tokio::main]
async fn main() {
    let (sender, receiver) = channel::<Message>(MESSAGE_BUFFER_SIZE);
    let manager = RuntimeHandlerManager::new ("xxx", sender).await.unwrap();

    let req = Request::CreateContainer(ContainerConfig {
            container_id: "xxx".to_owned(),
            bundle: "/tmp/bundle".to_owned(),
            rootfs_mounts: Vec::new(),
            terminal: false,
            options: None,
            stdin: None,
            stdout: None,
            stderr: None,
    });

    let result = manager.handler_message(req).await;
    println!("{:?}", result);
}

```
Despite being very simple and hacky (you can see I'm spoofing everything like there's no tomorrow) this code snippet ultimately allows me to properly execute `VirtSandbox::start()` which is the principal algorithm of kata VM startup.  I'm now able to call my `Qemu::prepare_vm()` and `Qemu::start_vm()` rather easily.

Granted, the `handler_message()` call above never returns at the moment because the sandbox startup tries to connect to the agent after VM launch which fails.  But this is just because my `prepare_vm()` and `start_vm()` currently do literally nothing apart from a `println!()` so it just means I'll have to gradually stop spoofing so much. ;-)

All in all I'm very hopeful about this as a road to a workable dev cycle.  If anybody's interested in details just let me know.  This could even be a way to implement the tool we're discussing here if we're comfortable to drive it at OCI level (I would be).

**Comment by jodh-intel** [comment]
Closing for now. The tool changes would be too invasive to `runtime-rs` and too unwieldy for developers.
